### DISCUSSION TITLE
error logging hooks

### DISCUSSION THREAD

**Discussion by hexsprite** [meteor/meteor#11528] [Feature Requests]
The Kadira APM client monkey patches several Meteor functions in order to capture exceptions in various places in the Meteor core.

It would much better if you could just register an error logger and it would be called for each error that was logged.

Areas that are currently patched include: Meteor._debug, Tracker exceptions, Blaze exceptions, and more I can't think of now.

Sketch of how this could be used:

```
import { registerErrorTracker } from 'meteor/logging'

registerErrorTracker((err) => sendToSentry(err))
```

This would easily allow production apps to log their errors to Sentry, Rollbar, Google Cloud Error tracking and others....

**Comment by StorytellerCZ** [comment]
I like this! Updating the logging package is one of the things that I want to do. I wanted to have this hook for all the logs called through it, in particular `Log.error`, but this goes a bit beyond that.

**Comment by filipenevola** [comment]
Hi @hexsprite this is a nice idea.

I believe the first step is simple, just add this new function to the logging package.

Of course for this to be useful we would need to adopt this pattern everywhere.

Am I missing something or the first step is straightforward?

**Reply by hexsprite** [reply]
Yeah, that sounds about right.

We could add a `logError()` and `registerErrorHandler()` to `meteor/logging` and start to transition the places that errors are logged in Meteor.

`logError` should probably accept an Error object, obviously, but also some context as that is often used in the tracker for example (Tracker error during afterFlush or recompute - see `_throwOrLog` and it's usages)

Another thing to consider is how many hoops packages currently use to capture errors logged used `Meteor._debug`, for example see:  https://github.com/monti-apm/monti-apm-agent/blob/master/lib/hijack/error.js#L84

**Reply by filipenevola** [reply]
We already [have](https://github.com/meteor/meteor/blob/6f0ecdafbde5ffb8652c98cd4660f4b91877cbe5/packages/logging/logging.js#L142-L144) [Log.error()](https://github.com/meteor/meteor/blob/a6c6bef4bee8535c9c4be67a3c0869fef471d482/packages/webapp/webapp_server.js#L558) why do we need logError?

Are you willing to start a PR adding `registerErrorHandler`?
